nix_update/version/github.py

```diff
@@ -34,3 +36,100 @@ def fetch_github_versions(url: ParseResult) -> List[Version]:
     tree = ET.fromstring(resp.read())
     releases = tree.findall(".//{http://www.w3.org/2005/Atom}entry")
     return [version_from_entry(x) for x in releases]
+
+
+def init_git(git_dir: str, url: str) -> None:
+    run(["git", "init", "-q", git_dir], stdout=None)
+    run(
+        [
+            "git",
+            "-C",
+            git_dir,
+            "remote",
+            "add",
+            "origin",
+            url,
+        ],
+        stdout=None,
+    )
+
+
+def get_latest_snapshot_from_git(
+    git_dir: str, branch: str, before: str, increment: int = 20, initial: int = 20
+) -> Version:
+    run(
+        [
+            "git",
+            "-C",
+            git_dir,
+            "fetch",
+            f"--depth={initial + increment}",
+            "origin",
+            branch,
+        ],
+        stdout=None,
+    )
+    while True:
+        log = run(
+            [
+                "git",
+                "-C",
+                git_dir,
+                "log",
+                "--format=%cd %H",
+                "--date=format-local:%F",
+                f"--before={before}T00:00:00Z",
+                "--max-count=1",
+                f"origin/{branch}",
+            ],
+            extra_env={"TZ": "UTC"},
+        ).stdout.strip()
+        if log:
+            date, commit = log.split(maxsplit=1)
+            return Version(f"unstable-{date}", rev=commit)
+        run(
+            [
+                "git",
+                "-C",
+                git_dir,
+                "fetch",
+                f"--deepen={increment}",
+                "origin",
+                branch,
+            ],
+            stdout=None,
+        )
+
+
+def fetch_github_snapshots(url: ParseResult, branch: str) -> List[Version]:
+    if url.netloc != "github.com":
+        return []
+    parts = url.path.split("/")
+    owner, repo = parts[1], parts[2]
+    repo = re.sub(r"\.git$", "", repo)
+    feed_url = f"https://github.com/{owner}/{repo}/commits/{branch}.atom"
+    info(f"fetch {feed_url}")
+    resp = urllib.request.urlopen(feed_url)
+    tree = ET.fromstring(resp.read())
+    commits = tree.findall(".//{http://www.w3.org/2005/Atom}entry")
+
+    before = datetime.utcnow().strftime("%Y-%m-%d")
+    snapshots = []
+    for entry in commits:
+        link = entry.find("{http://www.w3.org/2005/Atom}link")
+        updated = entry.find("{http://www.w3.org/2005/Atom}updated")
+        if link is None or updated is None or updated.text is None:
+            continue
```

Comment:
Since the data "should" be there and we just do it for the sake of mypy, this might be better for us.

```suggestion
        assert link is None or updated is None or updated.text is None
```